### PR TITLE
Separate script and layout messages, issue #8843

### PR DESCRIPTION
**Pull request by jkachmar**
Separated layout-specific messages to the constellation out from the `ScriptMsg` enum into a `LayoutMsg` enum within `script_traits/script_msg.rs`, addresses [#8843](https://github.com/servo/servo/issues/8843).

I initially tried to move `LayoutMsg` into `layout_traits/lib.rs`, but this introduced a cyclic dependency: `layout_traits` depends on `script_traits` for the `LayoutTaskFactory` implementation, and `script_traits/script_task.rs` now depends on `LayoutMsg` for new layout channels in `InitialScriptState` and `ScriptTask`.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8958)

### CODE REVIEW

File: components/compositing/constellation.rs [hunk 1]
```diff
@@ -366,6 +377,7 @@ impl<LTF: LayoutTaskFactory, STF: ScriptTaskFactory> Constellation<LTF, STF> {
                 id: pipeline_id,
                 parent_info: parent_info,
                 constellation_chan: self.script_sender.clone(),
+                layout_constellation_chan: self.layout_sender.clone(),
```
Comment (Manishearth) [inline components/compositing/constellation.rs]: Any reason for not calling this `layout_chan`?

File: components/script/dom/bindings/trace.rs [hunk 2]
```diff
@@ -306,6 +307,13 @@ impl JSTraceable for ConstellationChan<ScriptMsg> {
     }
 }
 
+impl JSTraceable for ConstellationChan<LayoutMsg> {
```
Comment (Manishearth) [inline components/script/dom/bindings/trace.rs]: I think we should just use `no_jsmanaged_fields!` on `ConstellationChan<T>`?

File: components/layout_traits/Cargo.toml [hunk 3]
```diff
@@ -22,6 +22,9 @@ path = "../net_traits"
 [dependencies.profile_traits]
 path = "../profile_traits"
 
+[dependencies.style_traits]
+path = "../style_traits"
```
Comment (Manishearth) [inline components/layout_traits/Cargo.toml]: Why is this necessary?

File: components/script/Cargo.toml [hunk 4]
```diff
@@ -27,6 +27,9 @@ path = "../net_traits"
 [dependencies.profile_traits]
 path = "../profile_traits"
 
+[dependencies.layout_traits]
+path = "../layout_traits"
```
Comment (Manishearth) [inline components/script/Cargo.toml]: Why is this necessary? LayoutMsg is defined in script_traits

**Comment by Manishearth** [comment]
@bors-servo delegate=KiChjang

Looks good to me (modulo nits) at a glance.

**Comment by KiChjang** [comment]
-S-awaiting-review -S-needs-rebase +S-needs-code-changes 

Great job so far! I've left comments in places where care must be taken when dealing with passing messages to the constellation.

The most important idea here is to recognize that the script objects should only send `ScriptMsg`s to the constellation - it should not, in anyway, need a `ConstellationChan<LayoutMsg>`. That's reserved for layout objects.

Secondly, both script and layout objects should never need to know what kind of messages (i.e. `ScriptMsg`and `LayoutMsg`) that the underlying channel accepts - all they need to know is that they are, in fact, communicating with the constellation, hence the `ScriptMsg as ConstellationMsg` in script files and `LayoutMsg as ConstellationMsg` in layout files.

---

Reviewed 11 of 19 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/compositing/constellation.rs, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dG8Iur1FtKJwaYQlq:-K5dG8Iur1FtKJwaYQlr:344707475) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/compositing/constellation.rs#L102)):*
nit: Should follow the comment style of previous receivers, i.e. "Receives messages from the layout task".

---

*[components/compositing/pipeline.rs, line 85 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dGxh1sQsq0HGF8nlB:-K5dGxh1sQsq0HGF8nlC:632019404) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/compositing/pipeline.rs#L85)):*
This is not actually true; it's a channel for the layout task to send messages to the constellation.

---

*[components/script/dom/document.rs, line 1163 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dJ9WsQFcQfVcbFohR:-K5dJ9WsQFcQfVcbFohS:1593766574) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/script/dom/document.rs#L1163)):*
This sounds very wrong to me. A document is a script object, and yet it is sending a LayoutMsg to the constellation. We should instead make two copies of the `ChangeRunningAnimationsState` message: one for `LayoutMsg`s, and the other for `ScriptMsg`s.

---

*[components/script/dom/window.rs, line 199 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dJiyOE6q3SO3KQ99E:-K5dJiyOE6q3SO3KQ99F:-1809972545) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/script/dom/window.rs#L199)):*
Again, wrong idea here. It's a channel to communicate with the constellation, but given the comment I left on document.rs, we really should remove this channel altogether.

---

*[components/script/layout_interface.rs, line 21 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dJzcbyMS4zjdRZbw7:-K5dJzcbyMS4zjdRZbw8:-126149560) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/script/layout_interface.rs#L21)):*
`LayoutMsg as ConstellationMsg`

---

*[components/script/lib.rs, line 51 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dK536SNFwNvTYMiXA:-K5dK536SNFwNvTYMiXB:987212923) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/script/lib.rs#L51)):*
Shouldn't be necessary anymore, since the original plan of moving `LayoutMsg` to `layout_traits` was scrapped.

---

*[components/script_traits/lib.rs, line 262 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8958#-K5dLKzEXsk6DUxlEQim:-K5dLKzEXsk6DUxlEQin:2056337305) ([raw file](https://github.com/servo/servo/blob/67dbf977a5fa9d34cb58b623b39e122b274768dd/components/script_traits/lib.rs#L262)):*
"from layout"; but we really should remove this, since this is from the script, and scripts should only ever send `ScriptMsg`s to the constellation.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8958#-:-K5dKwXQiQHerzkqqjCs:-1129403464)*

**Comment by KiChjang** [comment]
-S-awaiting-review -S-needs-rebase +S-needs-code-changes 

---

Reviewed 13 of 16 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/layout_task.rs, line 692 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8958#-K6C74xz0UYPk4x0EuTY:-K6C74xz0UYPk4x0EuTZ:-1627790553) ([raw file](https://github.com/servo/servo/blob/537f7d736299b6595c9e67fc1aa2d31b21d4b7f7/components/layout/layout_task.rs#L692)):*
Since there's no other type of `ConstellationChan`, we can just leave it named as `constellation_chan`.

---

*[components/script/layout_interface.rs, line 259 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8958#-K6C4HihmRJ8Eg-hJu0V:-K6C4HiiCda2dxk-AlTM:-403134690) ([raw file](https://github.com/servo/servo/blob/537f7d736299b6595c9e67fc1aa2d31b21d4b7f7/components/script/layout_interface.rs#L259)):*
I'm confused, why is there a `ConstellationChan<ScriptMsg>` here?

---

*[components/script/script_task.rs, line 1203 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8958#-K6C4hx1AvIB2DTT0Wzs:-K6C4hx1AvIB2DTT0Wzt:-836713124) ([raw file](https://github.com/servo/servo/blob/537f7d736299b6595c9e67fc1aa2d31b21d4b7f7/components/script/script_task.rs#L1203)):*
There's absolutely no need for the layout to have a `ConstellationChan<ScriptMsg>` at all. I don't find it necessary to add a new `layout_to_constellation_chan` field. You can just change the message type of `constellation_chan` to be `LayoutMsg`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8958#-:-K6C7fX6HniarRtwznXn:904062486)*

**Comment by KiChjang** [comment]
-S-awaiting-review 

@bors-servo r+

Thanks for doing this!

---

Reviewed 5 of 11 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8958#-:-K6VHEYNpTWoeVXIsiBp:-1801506476)*
